### ISSUE TITLE
Review feedback on MP JWT guide

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/guide-microprofile-jwt#108]
Some feedback and suggested updates based on recently learning about JWT from this guide and elsewhere and trying to understand what the guide is showing:
- The guide doesn't explain what each microservice does or where the JWTs get built. I've added a note to #34 because a diagram would help but also needs a brief description of the three microservices (see #34 for suggested details).
- Would it be useful to show the content of the token to make it clearer what information is in it? (and maybe link to jwt.io/introduction which has a nice introduction to what JWTs are)
- In"Try what you'll build"
  - "Note that you cannot directly visit a back end URL from a browser." - can you give the back-end URL so they can try it and see that it doesn't work (I just wanted to know for myself!). Then give the front-end URL so they can see what happens then.
- The guide uses approach of updating existing files, which is not the approach we take for the OL guides. It also isn't necessary here and makes the formatting of the 'action boxes' weird. In the two "Securing xxx service" sections, just say "Create the SystemResource class" etc and then have them create the class like they do in the other guides. Remove the instruction to "Add the `@RolesAllowed` annotation. Then reword the text afterwards to say something like "The SystemResource class is annotated with the `@RolesAllowed` annotation which provides roles-based access control to the class." etc. Can even say: "Without this annotation, the service can be accessed by any other service." to make clear what the difference is between adding and not adding the annotation.
- Also, if we can make this guide use the MP Rest client consistent with the MP Rest client guide (see #105 ), suggest at an appropriate point that the reader compares the non-secured code in that guide with the secured code in this guide.
- "Securing the inventory service" section
  - I think the heading "Adding the SecureSystemClient class" heading is not that informative and is in the wrong place - it needs to focus on the goal, not the implementation. I suggest rewording the heading to "Requesting information from the secured system service" (or similar) then moving the info about `getPropertiesForHost()` and `manager.get(hostname, authHeader)` down into that section (with a brief intro before it to give it context) (ie end the "Securing the inventory service" section after "…is properly secured, as well."). This will also fix my next point...
  - The first para makes sense - ie you've added roles-based access to the inventory service like you did for the system service. But then, it starts "In addition…" which seems to come from nowhere - it needs to introduce the concept of the security token being in the HTTP Authorization header etc
- "Trying more JAX-RS methods to validate tokens" - it's not clear why (there's no context) this section is here:
  - What is this section showing that the previous section couldn't show? It's not clear what the previous section showed that is now being built on to show something else. Is this section even necessary?
  - "To demonstrate…" seems the wrong wording. The author of the guide is demonstrating; the reader is "trying out" or "seeing" etc, not "demonstrating". eg "To see how MP JWT retrieves….".
  - This section is now introducing more new concepts without explaining them:
      - What are "claims" and what are "custom claims" (versus non-custom claims??)? Why do I need to care about them now? Are custom claims something I'm commonly going to need? - give an example real-life scenario of when. What is the custom claim in this code for?
      - What does "validating claims from security tokens" mean? Why do you need to do it?
      - Why do you need to show "how JAX-RS methods map to MP JWT features"? (what are "features" in this context?)
      - Why are getters and setters relevant here - why do claims need to be made available through getters?
      - It feels like this section is just describing what is technically happening (which the reader can already see from reading the code alongside) but not explaining why. Need to work out how to explain what's going on in logical steps and why the step matter/what it achieves, then highlight the bits of code that achieve this. But don't go into great detail (I know that makes it tricky in a guide) - we'll try to ensure there are conceptual topics in the OL Docs so that the guides can link out to more info.
    - Use of "JSON Web Token" instead of "JWT" is inconsistent in this section. If you just want to say "token" it's fine instead of "JSON Web Token" at this stage in the guide.
- "Building" section - Again, give the back-end URL so they can see what happens and that it doesn't work. Then give them the front-end URL to show how it does work.

I've had a couple of different people saying they find this guide difficult to understand/use. So I'm attempting to learn enough about the technology to see what might be confusing. This feedback is my first pass at that. There may be more as I understand it better. :)

**Comment by vsmid** [comment]
Hi, 

I ve been trying mpJwt feature of OL by just using microprofile-config.properties mp.jwt.verify.publickey and mp.jwt.verify.issuer (omitted mpJwt tag from server.xml). I generated jwt with Adam Bien's jwtenizr and that works great on Payara. 
I am using microprofile 3.0 feature with mpJwt 1.1 on OL 19.0.0.9.

However, on OL i keep getting:

CWWKS5523E: The MicroProfile JWT feature cannot authenticate the request because a valid JWT cannot be created from the token included in the request. CWWKS6029E: The JSON Web Token (JWT) cannot be validated because a signing key cannot be found. The configured signature algorithm [RS256] requires a key to validate the token.

I am not sure why it asks for private key. 

It would be great if guide would show an example of how to configure mp jwt without using server.xml mpJwt tag and keystores.

**Comment by vsmid** [comment]
Hi, 

just an update,

if i put the whole public key declaration like this it works:

mp.jwt.verify.publickey=-----BEGIN PUBLIC KEY-----MIIB...-----END PUBLIC KEY-----

Maybe this should be fixed or?

**Comment by chunlongliang-ibm** [comment]
agree we should have a guide for mpJwt-1.1.  The error message you see is misleading, I will investigate it. Could you paste your server.xml, and configured mpConfig for mpJwt-1.1?

mp.jwt.verify.publickey
    Specifies the embedded key material of the public key for the MP-JWT signer in PKCS8 PEM or JWK(S) format. If not found, check the mp.jwt.verify.publickey.location value.
mp.jwt.verify.publickey.location
    Specifies the relative path or full URL of the public key. All relative paths are resolved within the archive by using ClassLoader.getResource. If the value is a URL, the URL is resolved by using new URL(“...”).openStream().
mp.jwt.verify.issuer
    Specifies the expected iss claim value to validate against an MP-JWT.

**Comment by vsmid** [comment]
**server.xml**

```xml
<server description="Sample Liberty server">

  <featureManager>
    <feature>microProfile-3.0</feature>
  </featureManager>

  <applicationManager autoExpand="true" />

  <mpMetrics authentication="false" />

  <httpEndpoint host="*" httpPort="${default.http.port}" httpsPort="${default.https.port}" id="defaultHttpEndpoint"/>
  <keyStore id="defaultKeyStore" password="Liberty" />

  <webApplication location="io.openliberty.sample.getting.started.war" contextRoot="/"/>
  
</server>
```


**microprofile-config.properties**
```properties
#generated by jwtenizr
#Sat Oct 12 07:14:43 CEST 2019
mp.jwt.verify.publickey=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAmXtjPA0zH6H6LquhloPJbHyDW2Ahx8qaA9AubwiIZMd8c7gQ77npMqOSeZ5xvrnqqLa8EXUO4N8EldCsvGsepTDezjq5GIDRYmiVdY/Xn/E4WDp4CHvMqNXP8IhuBGTym5GLKgc5xIDIVF+8Z1tG0TQqMWhm/pGv4Ijq4FwNyp6gKmQa7WeNxN+XaYR773fznYRGRTu0BJ/QAGWC5EfdlQbHOpB/hyhBpV1IhzxyYZm3W0ZAsURlnENCZ16p4F6L1GmTKJxMTMUXQAjrSHtP7noKE5cN2pXmkX9LMk4l1c/Ek4ivRLvkOrgbBXIuc4jyN+Tn5WrigNG4DwvZYhL+YwIDAQAB
mp.jwt.verify.issuer=airhacks
```


When it works it looks like this:

**microprofile-config.properties**
```properties
#generated by jwtenizr
#Sat Oct 12 07:14:43 CEST 2019
mp.jwt.verify.publickey=-----BEGIN PUBLIC KEY-----MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAmXtjPA0zH6H6LquhloPJbHyDW2Ahx8qaA9AubwiIZMd8c7gQ77npMqOSeZ5xvrnqqLa8EXUO4N8EldCsvGsepTDezjq5GIDRYmiVdY/Xn/E4WDp4CHvMqNXP8IhuBGTym5GLKgc5xIDIVF+8Z1tG0TQqMWhm/pGv4Ijq4FwNyp6gKmQa7WeNxN+XaYR773fznYRGRTu0BJ/QAGWC5EfdlQbHOpB/hyhBpV1IhzxyYZm3W0ZAsURlnENCZ16p4F6L1GmTKJxMTMUXQAjrSHtP7noKE5cN2pXmkX9LMk4l1c/Ek4ivRLvkOrgbBXIuc4jyN+Tn5WrigNG4DwvZYhL+YwIDAQAB-----END PUBLIC KEY-----
mp.jwt.verify.issuer=airhacks
```

**Comment by chunlongliang-ibm** [comment]
The current open-liberty implementation strictly check signature verification key formats as described in spec. However, we would consider to relax the key format requirement.
https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/configuration.asciidoc#pcks8

**Comment by griffinhadfield** [comment]
This issue has been addressed as part of the guide's rewrite which can be seen on the `vertex-jwt-branch`
